### ISSUE TITLE
Add user callback for handling invalid characters.

### ISSUE THREAD

**Issue by ashtuchkin** [ashtuchkin/iconv-lite#53]
Inspired by ICU, something like:

``` javascript
  buf = iconv.encode(str, 'win1251', { invalidCharHandler: function(char) {
    // Here you can either throw exception which will be propagated outside,
    // or return a replacement char.
  });
```

Probably, some default handlers would also be nice (always throwing, always returning '?', trying to transliterate).

**Comment by peakji** [comment]
Any news on this? This would be a great feature to have.

**Comment by ashtuchkin** [comment]
I agree, I don't really have time to do it now. The main obstacle is that we need to reimplement all the native Node encodings from scratch to support this feature and it's not easy.

**Comment by bpasero** [comment]
Yes, would also be interested in this. I think this would enable VS Code to implement a dialog like this one:

![image](https://user-images.githubusercontent.com/900690/36909086-626bd878-1e3d-11e8-8900-4b18bc45a4fa.png)

Refs: https://github.com/Microsoft/vscode/issues/44505

**Comment by lapo-luchini** [comment]
> The main obstacle is that we need to reimplement all the native Node encodings from scratch to support this feature and it's not easy.

Maybe re-encoding to original charset and checking that input and regenerated string are equals?
I guess it should be faster than re-implementing that. (but truth be said, I didn't check the code)

**Comment by DecimalTurn** [comment]
> The main obstacle is that we need to reimplement all the native Node encodings from scratch to support this feature and it's not easy.

A simpler approach that I'm suggesting in #388 for single-byte encodings (SBCS) is to perform a check right before writting to the string buffer to see if we are about to write `?`/`�` for a character that was not `?`/`�` in the original string. That approach should be applicable to multi-byte encodings, but I just wanted to validate the implementation for SBCS before going further.
